README
$Id: README,v 1.1 2003-02-08 13:30:30 eric Exp $

--------
Abstract
--------

This module subclasses Log::Dispatch::File for logging to date/time
stamped files.

------------
Requirements
------------

This module requires Log::Dispatch::File, included in the Log::Dispatch
distribution.

------------------
Basic Installation
------------------

Log::Dispatch::File::Stamped may be installed through the CPAN shell in the
usual manner:

  # perl -MCPAN -e 'install Log::Dispatch::File::Stamped'

You can also read this README from the CPAN shell:

  # perl -MCPAN -e shell
  cpan> readme Log::Dispatch::File::Stamped

And you can install the component from the CPAN prompt as well:

  cpan> install Log::Dispatch::File::Stamped

-------------------
Manual Installation
-------------------

Log::Dispatch::File::Stamped can also be installed manually.  The latest CPAN
version can be found at
<http://www.cpan.org/modules/by-authors/id/C/CH/CHOLET/> or in a
similarly named directory at your favorite CPAN mirror.

Downloading and unpacking the distribution are left as exercises for
the reader.  To build and test it:

  perl Makefile.PL
  make test

When you're ready to install the component:

  make install

It should now be ready to use.

On Win32 systems, replace "make" in the above commands with "nmake".
The nmake utility can be downloaded from
C<http://download.microsoft.com/download/vc15/Patch/1.52/W95/EN-US/Nmake15.exe>


Changes
Revision history for Log-Dispatch-File-Stamped

{{$NEXT}}

0.21      2022-10-23 02:19:49Z
          - fix Specio prereq (RT#144903)

0.20      2022-10-22 18:27:55Z
          - new config value "stamp_sep", for changing the separator character
            in the log filename (PR #2, thanks Sven Willenbuecher!)

0.19      2017-02-21 03:59:32Z
          - fix unit tests for different path separator on MSWin32

0.18      2017-02-18 19:21:37Z
          - fix checking of close_after_write option that changed internally
            in Log::Dispatch 2.59

0.17      2017-02-18 06:25:35Z
          - fix test that changed behaviour with Log::Dispatch 2.59
          - new 'time_function' argument, for optionally using gmtime instead
            of localtime for generating the filename timestamp (see RT#113276)
          - additional internal caching layer added, to avoid unnecessary
            system calls when several messages are logged at once (thanks to
            Doug Hoyte for the suggestion; see RT#113275)

0.16      2016-01-28 01:25:59Z
          - fix close-after-write tests that broke with Log::Dispatch::File
            2.53, which now deletes the stored filehandle after closing
            (RT#111292)

0.15      2014-11-05 02:28:41Z
          - fix tests on MSWin32 using arch-sensitive file separators

0.14      2014-11-03 06:14:34Z
          - fix lingering Test::TempDir prereq
          - fix remaining issues with running tests in parallel by using
            unique tempdirs for all tests

0.13      2014-02-17 01:14:47Z
          - uses of Test::TempDir switched to Path::Tiny

0.12      2013-06-17 21:57:03Z
          - conversion to Dist::Zilla

0.11      2013-06-17
          - re-release as a stable release

0.10_010  2013-05-04
          - fixed testing of octal values on MSWin32, ONCE AND FOR ALL

0.10_009  2013-05-01
          - even more fixing of testing of octal values on MSWin32

0.10_008  2013-04-30
          - more fixing of testing of octal values on MSWin32

0.10_007  2013-04-29
          - fix testing of octal values on MSWin32

0.10_003  2013-04-22
          - fix version declaration (Karen Etheridge)

0.10_002  2013-04-22
          - fix version requirements (Karen Etheridge)

0.10_001  2013-04-09
          - add support for all remaining Log::Dispatch::File options (Karen Etheridge)

0.10  2011-09-25
          - add support for permissions parameter - reported by Jonathan Peyser

0.09  2010-08-15
          - fix incorrect number of skipped tests under perl < 5.8

0.08  2010-08-14
          - skip binmode tests under perls < 5.8

0.07  2010-08-13
          - add support for binmode parameter - patch by Tokuhiro Matsuno
            (https://rt.cpan.org/Ticket/Display.html?id=60278)

0.06  2008-06-29
          - added missing meta-spec to meta.yml
          - remove pod tests

0.05  2006-09-07
          - fix typo which prevented correcting indexing by PAUSE
          - move .pm file to lib/
          - document log_message()

0.04  2006-09-07
          - add Build.PL to build with Module::Build
          - add t/podcover.t to make CPANTS happy

0.03  2004-09-15
          - Fix "Log::Dispatch::* object file already exists" warning
            displayed on some platforms during "make test".
          - Require Test::More
          - Fix pod and add a pod test

0.02  2004-08-08
          - log_to() wasn't working due to attribute name clash
            (https://rt.cpan.org/Ticket/Display.html?id=7249)

0.01  2003-02-08
          - Uploaded 0.01 to CPAN



lib/Log/Dispatch/File/Stamped.pm
use strict;
use warnings;
package Log::Dispatch::File::Stamped;
# vim: set ts=8 sts=4 sw=4 tw=115 et :
# ABSTRACT: Logging to date/time stamped files
# KEYWORDS: log logging output file timestamp date rolling rotate

our $VERSION = '0.22';

use File::Basename        qw(fileparse);
use File::Spec::Functions qw(catfile);
use POSIX                 qw(strftime);

use Log::Dispatch::File 2.38;
use parent 'Log::Dispatch::File';

use Params::ValidationCompiler qw(validation_for);

use Specio::Library::Builtins;
use Specio::Library::String;
use Specio::Declare 0.48;

use namespace::clean 0.19;

enum(
    'TimeFunctions',
    values => [qw(localtime gmtime)],
);

sub _basic_init
{
    my $self = shift;

    $self->SUPER::_basic_init(@_);

    my $validator = validation_for(
        params => {
            stamp_fmt => {
                type => t('SimpleStr'),
                default => '%Y%m%d',
            },
            stamp_sep => {
                type => t('SimpleStr'),
                default => '-',
            },
            time_function => {
                type => t('TimeFunctions'),
                default => 'localtime',
            },
        },
        slurpy => 1,
    );
    my %p = $validator->(@_);

    $self->{stamp_fmt} = $p{stamp_fmt};
    $self->{stamp_sep} = $p{stamp_sep};
    $self->{time_function} = $p{time_function};

    # cache of last timestamp used
    $self->{_stamp} = '';
    $self->{_time} = 0;

    # split pathname into basename, path, extension
    @$self{qw(_name _path _ext)} = fileparse($self->{filename}, '\.[^.]+');

    # stored in $self->{filename} (overwrites original); used by _open_file()
    $self->_make_filename;
}

sub _make_filename
{
    my $self = shift;

    # re-use last filename if the time has not changed
    my $time = time();
    return $self->{filename} if $time eq $self->{_time};

    $self->{_time} = $time;
    my $stamp = strftime($self->{stamp_fmt}, $self->{time_function} eq 'localtime' ? localtime : gmtime);

    # re-use last filename if the stamp has not changed
    return $self->{filename} if $stamp eq $self->{_stamp};

    $self->{filename} = catfile(
        $self->{_path},
        join($self->{stamp_sep}, $self->{_name}, $stamp) . ($self->{_ext} ? $self->{_ext} : '')
    );
}

sub log_message
{
    my $self = shift;

    # check if the filename is the same as last time...
    my $old_filename = $self->{filename};
    $self->_make_filename;

    # don't re-open if we use close-after-write - the superclass will do it
    if (not $self->{ Log::Dispatch->VERSION >= '2.59' ? 'close_after_write' : 'close' }
            and $old_filename ne $self->{filename})
    {
        $self->_open_file;
    }

    $self->SUPER::log_message(@_);
}

1;
__END__

=pod

=head1 SYNOPSIS

  use Log::Dispatch::File::Stamped;

  my $file = Log::Dispatch::File::Stamped->new(
    name      => 'file1',
    min_level => 'info',
    filename  => 'Somefile.log',
    stamp_fmt => '%Y%m%d',
    stamp_sep => ':',
    time_function => 'localtime',
    mode      => 'append',
  );

  $file->log( level => 'emerg', message => "I've fallen and I can't get up\n" );

=head1 DESCRIPTION

This module subclasses L<Log::Dispatch::File> for logging to date/time-stamped
files, respecting all its configuration options. As with other L<Log::Dispatch>
handlers, the destination file is kept open for as long as the filename remains
constant (unless C<close_after_write> is set).

=head1 METHODS

=head2 new(%p)

This method takes the same set of parameters as L<Log::Dispatch::File::new()|Log::Dispatch::File/new>,
with the following differences:

=over 4

=item * filename ($)

The filename template. The actual timestamp will be appended to this filename
when creating the actual logfile. If the filename has an extension, the
timestamp is inserted before the extension. See examples below.

=item * stamp_fmt ($)

The format of the timestamp string. This module uses L<POSIX::strftime|POSIX/strftime> to
create the timestamp string from the current local date and time.
Refer to your platform's C<strftime> documentation for the list of allowed
tokens.

Defaults to C<%Y%m%d>.

=item * stamp_sep ($)

The separator character(s) used between components in the log filename.

Defaults to C<->.

=item * time_function ($)

The function used to determine the current time; present choices are L<perlfunc/localtime> or
L<perlfunc/gmtime>. Defaults to C<localtime>.

=back

=head2 log_message( message => $ )

Sends a message to the appropriate output.  Generally this
shouldn't be called directly but should be called through the
C<log()> method (in L<Log::Dispatch::Output>).

=head1 EXAMPLES

=for stopwords txt

Assuming the current date and time is:

  % perl -e 'print scalar localtime'
  Sat Feb  8 13:56:13 2003

  Log::Dispatch::File::Stamped->new(
    name      => 'file',
    min_level => 'debug',
    filename  => 'logfile.txt',
  );

This will log to file F<logfile-20030208.txt>.

  Log::Dispatch::File::Stamped->new(
    name      => 'file',
    min_level => 'debug',
    filename  => 'logfile.txt',
    stamp_fmt => '%d%H',
  );

This will log to file F<logfile-0813.txt>.

=head1 SEE ALSO

=for :list
* L<Log::Dispatch::File>
* L<POSIX>
* L<Log::Dispatch::File::Rolling>
* L<Log::Dispatch::FileRotate>
* L<Log::Dispatch::FileWriteRotate>

=head1 ACKNOWLEDGEMENTS

=for stopwords Rolsky

Dave Rolsky, author of the L<Log::Dispatch> suite and many other
fine modules on CPAN.

This module was rewritten to respect all present (and future) options to
L<Log::Dispatch::File> by Karen Etheridge, <ether@cpan.org>.

=cut


Makefile.PL
# $Id: Makefile.PL,v 1.3 2006-09-07 08:40:27 eric Exp $

use ExtUtils::MakeMaker 6.30;

WriteMakefile(
    NAME          => 'Log::Dispatch::File::Stamped',
    AUTHOR        => 'Eric Cholet <cholet@logilune.com>, Karen Etheridge <ether@cpan.org>',
    VERSION_FROM  => 'lib/Log/Dispatch/File/Stamped.pm',
    ABSTRACT_FROM => 'lib/Log/Dispatch/File/Stamped.pm',
    PL_FILES      => {},
    PREREQ_PM     => {
                      'Log::Dispatch'       => 0,
                      'Log::Dispatch::File' => '2.38',
                      'Test::More'          => '0.88',
                      'Test::TempDir'       => 0,
                      'Path::Tiny'          => 0,
                    },
    dist          => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean         => { FILES => 'Log-Dispatch-File-Stamped-*' },

    META_MERGE => {
        dynamic_config => 0,

        resources => {
            bugtracker => 'https://rt.cpan.org/Public/Dist/Display.html?Name=Log-Dispatch-File-Stamped',
            repository => 'https://github.com/karenetheridge/Log-Dispatch-File-Stamped.git',
            homepage => 'https://github.com/karenetheridge/Log-Dispatch-File-Stamped',
        },
    },
);
